src/index.js

```diff
@@ -2,56 +2,147 @@ import F from 'futil'
-      // and need for mobx 4
-      value: _.getOr('', 'value', x),
-      get errors() {
-        return form.errors[field] || []
-      },
       get isValid() {
-        return !node.errors.length
+        return _.isEmpty(node.errors)
       },
       get isDirty() {
-        return changed(toJS(node.value), toJS(x.value))
+        return changed(
+          toJS(getOrValue(path, baseNode.value)),
+          toJS(getOrValue(path, form.value))
+        )
       },
+      getField: x => _.get(joinPaths(buildFieldPath(x)), node.fields),
       reset() {
-        node.value = F.when(_.isUndefined, '')(x.value)
+        if (_.isEmpty(path)) form.submit.state.error = null // Lil hack
+        node.value = node.errors = node.fields = undefined
+        let baseValue = getOrValue(path, baseNode.value)
+        let baseField = _.getOr(baseNode, buildFieldPath(path), baseNode.fields)
+        if (baseField.items) {
+          node.add(baseValue)
+        } else if (baseField.fields) {
+          node.add(
+            F.mapValuesIndexed(
+              (x, k) => F.compactObject({ ...x, value: _.get(k, baseValue) }),
+              baseField.fields
+            )
+          )
+        } else {
+          node.value = baseValue
+        }
+      },
+      validate(fields) {
+        let picked = pickFields(form, fields || _.join('.', path))
+        form.errors = {
+          ..._.omit(_.keys(picked), form.errors),
+          ...validate(form, picked),
+        }
+        return form.errors
       },
-      validate: () => form.validate([field]),
       clean() {
-        x.value = node.value
+        node.errors = undefined
+        F.setOn(['value', ...path], toJS(node.value), baseNode)
+        F.setOn(['fields', ...path], toJS(node.fields), baseNode)
       },
-      empty() {
-        if (_.isArray(toJS(x.value))) node.value = []
-        else if (_.isPlainObject(toJS(x.value))) node.value = {}
-        else node.value = ''
+      add(x) {
+        if (node.items) {
+          F.defaultsOn({ value: observable([]), fields: observable([]) }, node)
+          let lengthBefore = node.value.length
+          node.value.push(...observable(_.castArray(x)))
+          _.each(
+            index =>
+              node.fields.push(
+                initField(
+                  { ...node.items, field: index, value: node.value[index] },
+                  [...path, index]
+                )
+              ),
+            _.range(lengthBefore, node.value.length)
+          )
+        } else {
+          F.defaultsOn({ fields: observable({}), value: observable({}) }, node)
+          extendObservable(node.value, F.compactObject(_.mapValues('value', x)))
+          extendObservable(node.fields, initFields(x, path))
+        }
       },
-      ..._.omit('value', x),
+      remove(x) {
+        let parentPath = tokenizePath(x)
+        let path = parentPath.splice(parentPath.length - 1)
```

Comment:
`splitAt(-1)` or something 😄

Reply:
Actually I just discovered `chunk` from lodash so this can be done like
```
let tokens = tokenizePath(x)
let [parentPath, path] = _.chunk(tokens.length - 1, tokens)
```